### PR TITLE
[FLINK-27989][table-planner] Csv format supports reporting statistics

### PR DESCRIPTION
**Pull request by swuferhong**
## What is the purpose of the change

In FLIP-231, [https://cwiki.apache.org/confluence/display/FLINK/FLIP-231%3A+Introduce+SupportsStatisticReport+to+support+reporting+statistics+from+source+connectors](url), we aims to support csv, parquet, orc format for reporting statistics. This pr is a part of it to support CSV format. The details are as follows:

- In `CsvFileFormatFactory`,  for `CsvBulkDecodingFormat`, we implements `FileBasedStatisticsReportableInputFormat`  interface, and implements `reportStatistics()` method.
- Adding two UT tests, `CsvFormatStatisticsReportTest` and `CsvFormatFilesystemStatisticsReportTest` to test results.

## Brief change log

- In `CsvFileFormatFactory`,  for `CsvBulkDecodingFormat`, we implements `FileBasedStatisticsReportableInputFormat`  interface, and implements `reportStatistics()` method.
- Adding two UT tests.
- Modify `TestCsvFileFormatFactory` to make the result consistent in test method.

## Verifying this change

- Adding UT test `CsvFormatStatisticsReportTest` to test csv format only.
- Adding UT test `CsvFormatFilesystemStatisticsReportTest` to test csv format in filesystem connector.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector:  no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented?  no

### CODE REVIEW

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java [hunk 1]
```diff
@@ -136,6 +147,45 @@ public BulkFormat<RowData, FileSourceSplit> createRuntimeDecoder(
...
+                    if (sampledLineCnt < totalSampleLineCnt) {
+                        try (InputStreamReader isr =
+                                new InputStreamReader(
+                                        Files.newInputStream(new File(file.toUri()).toPath()))) {
+                            BufferedReader br = new BufferedReader(isr);
+                            String line;
+                            while (sampledLineCnt < totalSampleLineCnt
+                                    && (line = br.readLine()) != null) {
+                                sampledLineCnt += 1;
+                                sampledTotalSize += line.length();
```
Comment (fsk119) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java:173]: Why use line.length() here? I think it should be 

```
line.getBytes(StandardCharsets.UTF_8).length
```

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java [hunk 2]
```diff
@@ -136,6 +147,45 @@ public BulkFormat<RowData, FileSourceSplit> createRuntimeDecoder(
...
+                        }
+                    }
+                }
+                if (sampledTotalSize == 0) {
+                    return TableStats.UNKNOWN;
+                }
+
+                int realSampledLineCnt = Math.min(totalSampleLineCnt, sampledLineCnt);
+                int estimatedRowCount = (int) (totalSize * realSampledLineCnt / sampledTotalSize);
+                return new TableStats(estimatedRowCount);
```
Comment (fsk119) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java:184]: Why the estimation doesn't give the column stat?

File: flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFilesystemStatisticsReportTest.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    @Test
+    public void testCsvFileSystemStatisticsReport() {
+        FlinkStatistic statistic = getStatisticsFromOptimizedPlan("select * from sourceTable");
+        assertThat(statistic.getTableStats()).isEqualTo(new TableStats(3));
+    }
+
+    private void writeData(File file, List<String> data) throws IOException {
+        Files.write(file.toPath(), String.join("\n", data).getBytes());
```
Comment (fsk119) [inline flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFilesystemStatisticsReportTest.java:78]: In windows system, it uses "\r\n" as line feed.

File: flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFilesystemStatisticsReportTest.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Test for statistics functionality in {@link CsvFormatFactory}. */
+public class CsvFilesystemStatisticsReportTest extends TableTestBase {
```
Comment (fsk119) [inline flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFilesystemStatisticsReportTest.java:46]: Could you add some bad cases:

- The file doesn't exist.
- There are some dirty files in the directory. 
- ...

File: flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFilesystemStatisticsReportTest.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    private TableEnvironment tEnv;
+
+    @Before
+    public void setup() throws IOException {
+        util = batchTestUtil(TableConfig.getDefault());
+        tEnv = util.getTableEnv();
+        String path1 = tempFolder().newFile().getAbsolutePath();
+        writeData(new File(path1), Arrays.asList("1,1,hi", "2,1,hello", "3,2,hello world"));
+
+        String ddl1 =
```
Comment (fsk119) [inline flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFilesystemStatisticsReportTest.java:57]: ddl1 -> ddl

File: flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFilesystemStatisticsReportTest.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        @Override
+        public void visit(RelNode node, int ordinal, RelNode parent) {
+            if (node instanceof TableScan) {
+                Preconditions.checkArgument(result == null);
+                TableSourceTable table = (TableSourceTable) node.getTable();
+                result = table.getStatistic();
+            }
+            super.visit(node, ordinal, parent);
+        }
+    }
```
Comment (fsk119) [inline flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFilesystemStatisticsReportTest.java:101]: Why not just use the Reporter to get the statistic?

**Review by fsk119 — COMMENTED** [review]
Thanks for your contribution. Please add more detailed descrption.

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java [hunk 7]
```diff
@@ -136,6 +147,45 @@ public BulkFormat<RowData, FileSourceSplit> createRuntimeDecoder(
         public ChangelogMode getChangelogMode() {
             return ChangelogMode.insertOnly();
         }
+
+        @Override
+        public TableStats reportStatistics(List<Path> files, DataType producedDataType) {
+            final int totalSampleLineCnt = 100;
```
[the remark this answers is not in the document]
Comment (swuferhong) [reply flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java:153]: > 

Thanks @fsk119 . Not likes Orc and Parquet format,  Csv table stats are very rough estimate values, and don't include getting column statistics. Therefore, I think this value has little impact on the accuracy of Csv stats results.

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java (hunk 2 above)
Comment (swuferhong) [reply flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java:184]: > Why the estimation doesn't give the column stat?

Not likes Orc and Parquet format, Csv format not support collect stats. The only way to collect column stats is to build a csv reader and read all files which is a heavy operation.

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java [hunk 8]
```diff
@@ -136,6 +147,45 @@ public BulkFormat<RowData, FileSourceSplit> createRuntimeDecoder(
...
+            try {
+                long totalSize = 0;
+                int sampledLineCnt = 0;
+                long sampledTotalSize = 0;
```
[the remark this answers is not in the document]
Comment (swuferhong) [reply flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java:157]: > totalSize -> totalFileSize sampledLineCnt -> sampledRowCnt sampledTotalSize -> sampledRowSize

done！

File: flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFilesystemStatisticsReportTest.java (hunk 3 above)
Comment (swuferhong) [reply flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFilesystemStatisticsReportTest.java:78]: > In windows system, it uses "\r\n" as line feed.

this situation test on Unit Test.

File: flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFilesystemStatisticsReportTest.java (hunk 6 above)
Comment (swuferhong) [reply flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFilesystemStatisticsReportTest.java:101]: > Why not just use the Reporter to get the statistic?

This method is designed to reuse while extension for other source. This part will be abstracted during extension.

**Comment by swuferhong** [comment]
@flinkbot run azure

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java [hunk 9]
```diff
@@ -136,6 +150,52 @@ public BulkFormat<RowData, FileSourceSplit> createRuntimeDecoder(
         public ChangelogMode getChangelogMode() {
             return ChangelogMode.insertOnly();
         }
+
+        @Override
+        public TableStats reportStatistics(List<Path> files, DataType producedDataType) {
+            final int totalSampleLineCnt = 100;
```
Comment (godfreyhe) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java:156]: add some comment to explain why we use the magic number 100 here

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java [hunk 10]
```diff
@@ -136,6 +150,52 @@ public BulkFormat<RowData, FileSourceSplit> createRuntimeDecoder(
...
+                        try (InputStreamReader isr =
+                                        new InputStreamReader(
+                                                Files.newInputStream(
+                                                        new File(file.toUri()).toPath()));
+                                BufferedReader br = new BufferedReader(isr)) {
+                            String line;
+                            while (sampledRowCnt < totalSampleLineCnt
+                                    && (line = br.readLine()) != null) {
+                                sampledRowCnt += 1;
+                                sampledRowSize += line.getBytes(StandardCharsets.UTF_8).length;
```
Comment (godfreyhe) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java:177]: please update the TestCsvInputFormat

File: flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFormatStatisticsReportTest.java [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Test for statistics functionality in {@link CsvFormatFactory}. */
+public class CsvFormatStatisticsReportTest {
```
Comment (godfreyhe) [inline flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFormatStatisticsReportTest.java:40]: please add some tests which input files have more than 100 lines

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution, I left some comments

File: flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFormatStatisticsReportTest.java (hunk 11 above)
Comment (swuferhong) [reply flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFormatStatisticsReportTest.java:40]: > please add some tests which input files have more than 100 lines

done!

**Comment by swuferhong** [comment]
@flinkbot run azure

**Comment by swuferhong** [comment]
@flinkbot run azure

**Comment by swuferhong** [comment]
@flinkbot run azure

**Review by godfreyhe — APPROVED** [review]
LGTM

**Comment by godfreyhe** [comment]
@flinkbot run azure

**Comment by swuferhong** [comment]
@flinkbot run azure

**Comment by swuferhong** [comment]
@flinkbot run azure

**Comment by swuferhong** [comment]
@godfreyhe CI passed, could you help me to merge it ! Thanks!

**Pull request closed without merging**